userspace/lib/usbip_forward.c

```diff
@@ -607,6 +607,9 @@ usbip_forward(HANDLE hdev_src, HANDLE hdev_dst, BOOL inbound)
 	if (buff_dst.in_reading)
 		CancelIoEx(hdev_dst, &buff_dst.ovs[0]);
 
+	if (buff_src.in_reading || buff_dst.in_reading)
+		SleepEx(500, TRUE);
```

Comment:
Is 500msec sufficiently large? In almost case, a simple sleep does not solve a problem completely. It's a complicated point that CancelIoEx does not guarantee an async  I/O has been completed. How about using CancelIo ? Or exiting a sleep loop only after checking the done of overlapped structure.

Reply:
I attempted CancelIo and seemed to have the same issue. I think the issue is when we release our thread, the callback is able to run but since the buffers don't exist anymore we crash. 

I think any amount of sleep should allow the callback to run as we are allowing a yield. My feeling is that if it doesn't clean up in 500ms then we will essentially wait forever. If we wait forever we are leaking a thread and all of its memory forever. In that circumstance we should crash the application. To summarize I think 500ms is probably fine. Might be too long might be too short. I'm not really sure.

Reply:
@rpasek : How about considering the sample in [this page](https://docs.microsoft.com/en-us/windows/win32/fileio/canceling-pending-i-o-operations) ?

Reply:
Thanks for the info. I gave it a shot with this code as an alternative to the code in this commit:
```
	if (buff_src.in_reading) {
		CancelIoEx(hdev_src, &buff_src.ovs[0]);
		GetOverlappedResult(hdev_src, &buff_src.ovs[0], &num_bytes_discarded, TRUE);
	}
	if (buff_dst.in_reading) {
		CancelIoEx(hdev_dst, &buff_dst.ovs[0]);
		GetOverlappedResult(hdev_dst, &buff_dst.ovs[0], &num_bytes_discarded, TRUE);
	}
```

I tried a few variations of this code however the code still crashes on read_completion() when rbuff is accessed. I'm not sure why Microsoft's recommended solution doesn't work. Maybe our driver doesn't support cancelation? 

The sleep idea still seems to function in the circumstances that I've tested. Would you and @Oxalin like to test and see if it works for you as well? If so maybe we merge this work around for now?

Reply:
@rpasek : I have tested with GetOverlappedResultEx but the function failed returning ERROR_INVALID_HANDLE. This is due to that hEvent in OVERLAPPED structure is not an event object. Thus, checking in_reading repeatedly with your suggested sleep would be better.

```
diff --git a/userspace/lib/usbip_forward.c b/userspace/lib/usbip_forward.c
index 969cd75..a0e35a1 100644
--- a/userspace/lib/usbip_forward.c
+++ b/userspace/lib/usbip_forward.c
@@ -602,13 +602,15 @@ usbip_forward(HANDLE hdev_src, HANDLE hdev_dst, BOOL inbound)
 
 	/* Cancel an uncompleted asynchronous read */
 	/* If there's no asynchronous read pending, CancelIo seems to be blocked. */
-	if (buff_src.in_reading)
+	/* in_reading should be checked as cleared to guarantee that an IO completion routine has been called */
+	while (buff_src.in_reading) {
 		CancelIoEx(hdev_src, &buff_src.ovs[0]);
-	if (buff_dst.in_reading)
+		SleepEx(500, TRUE);
+	}
+	while (buff_dst.in_reading) {
 		CancelIoEx(hdev_dst, &buff_dst.ovs[0]);
-
-	if (buff_src.in_reading || buff_dst.in_reading)
 		SleepEx(500, TRUE);
+	}
 
 	cleanup_devbuf(&buff_src);
 	cleanup_devbuf(&buff_dst);
```

Reply:
@rpasek : What do you think about my suggestion?

Reply:
Sorry about the delay. I think this is fine. Do I need to modify anything in my PR?

Reply:
Thanks @rpasek . 
> Do I need to modify anything in my PR?

Yes, please merge my suggested patch into your commit. A squashed single commit will be easy to be merged into the master branch. If you have no time, I'll do.